lib/overcommit/hook_context/pre_push.rb

```diff
@@ -18,17 +18,14 @@ def pushed_refs
     end
 
     def modified_files
-      @modified_files ||= Overcommit::GitRepo.modified_files(refs: ref_range)
+      @modified_files ||= pushed_refs.map(&:modified_files).flatten
     end
 
     def modified_lines_in_file(file)
       @modified_lines ||= {}
-      @modified_lines[file] =
-        Overcommit::GitRepo.extract_modified_lines(file, refs: ref_range)
-    end
-
-    def ref_range
-      "#{pushed_refs[0].remote_sha1}..#{pushed_refs[0].local_sha1}"
+      @modified_lines[file] = pushed_refs.map do |pushed_ref|
+        pushed_ref.modified_lines_in_file(file)
+      end.to_set.flatten
```

Comment:
Can you help me understand why you are using `to_set.flatten` here?

`modified_lines_in_file` returns a `Set`, which you map over to produce an array of sets that you then call `to_set` on, followed by `flatten`. This can be potentially expensive for a large number of files and/or modifies lines. 

Consider the following benchmark:

```ruby
require 'benchmark'
require 'set'

iterations = 10
sets = 100.times.map do
  Array.new(1_000) { rand(0..1_000) }.to_set
end

Benchmark.bmbm do |x|
  x.report('.to_set.flatten') do
    iterations.times do
      sets.to_set.flatten
    end
  end

  x.report('each_with_object{ merge }') do
    iterations.times do
      sets.each_with_object(Set.new) { |item, set| set.merge(item) }
    end
  end
end

puts 'Are they equivalent?'
puts sets.to_set.flatten == sets.each_with_object(Set.new) { |item, set| set.merge(item) }
```

```
Rehearsal -------------------------------------------------------------
.to_set.flatten             0.340000   0.000000   0.340000 (  0.344824)
each_with_object{ merge }   0.020000   0.000000   0.020000 (  0.016745)
---------------------------------------------------- total: 0.360000sec

                                user     system      total        real
.to_set.flatten             0.360000   0.000000   0.360000 (  0.361935)
each_with_object{ merge }   0.020000   0.000000   0.020000 (  0.016793)
Are they equivalent?
true
```

In a nutshell, it would be significantly faster for large sets of files/lines if we didn't use `to_set.flatten`, but I want to make sure I understand your reasoning here in case I missed something.

Reply:
`flatten` is called because I assume the expected output is non-nested `Set` for `modified_lines_in_file` and non-nested `Array` for `modified_files`. 

For example, `modified_lines_in_files` with 2 sets on `PushedRef` (when I pushed 2 branches) might output `[<Set(1,2)>, <Set(3,4)>]‘. Then calling`to_set.flatten`, converts it to `<Set(1,2,3,4)>`. 

I definitely will go with your suggestion if the output is the same since it has better performance.